### PR TITLE
Get sql from audit logs

### LINKED ISSUE #1694: Implement RFC 006 - Generate SQL from audit table

### PR DESCRIPTION
**Pull request by aguillenv**
RFC 006

Resolves #1694 

Query examples:
```sql
SELECT * FROM iasql_get_sql_since();
SELECT * FROM iasql_get_sql_since('2023-01-05T12:00:00');
SELECT * FROM iasql_get_sql_since((now() - interval '5 minutes')::text);
```

Result examples:
```sql
INSERT INTO load_balancer (load_balancer_name, load_balancer_arn, dns_name, canonical_hosted_zone_id, created_time, scheme, state, load_balancer_type, subnets, availability_zones, ip_address_type, region, attributes, vpc)
VALUES ('test', 'arn:aws:elasticloadbalancing:us-east-2:257682470237:loadbalancer/app/test/88a5af6043314c2b', 'test-506077814.us-east-2.elb.amazonaws.com', 'Z3AADJGX6KTTL2', '2023-01-05T16:49:16.99+00:00', 'internet-facing', 'active', 'application', array['subnet-086d1c2cc46bdd254','subnet-0bfe61253a129d77d']::varchar[], array['us-east-2a','us-east-2b']::varchar[], 'ipv4', (SELECT region FROM aws_regions WHERE region = 'us-east-2'), '[{"Key":"access_logs.s3.enabled","Value":"false"},{"Key":"access_logs.s3.bucket","Value":""},{"Key":"access_logs.s3.prefix","Value":""},{"Key":"idle_timeout.timeout_seconds","Value":"60"},{"Key":"deletion_protection.enabled","Value":"false"},{"Key":"routing.http2.enabled","Value":"true"},{"Key":"routing.http.drop_invalid_header_fields.enabled","Value":"false"},{"Key":"routing.http.xff_client_port.enabled","Value":"false"},{"Key":"routing.http.preserve_host_header.enabled","Value":"false"},{"Key":"routing.http.xff_header_processing.mode","Value":"append"},{"Key":"load_balancing.cross_zone.enabled","Value":"true"},{"Key":"routing.http.desync_mitigation_mode","Value":"defensive"},{"Key":"waf.fail_open.enabled","Value":"false"},{"Key":"routing.http.x_amzn_tls_version_and_cipher_suite.enabled","Value":"false"}]'::json, (SELECT id FROM vpc WHERE vpc_id = 'vpc-024bfb31f9340a7cc' AND region = (SELECT region FROM aws_regions WHERE region = 'us-east-2')));

INSERT INTO load_balancer_security_groups (load_balancer_id, security_group_id)
VALUES ((SELECT id FROM load_balancer WHERE load_balancer_arn = 'arn:aws:elasticloadbalancing:us-east-2:257682470237:loadbalancer/app/test/88a5af6043314c2b'), (SELECT id FROM security_group WHERE group_id = 'sg-06204058f6faf6f52' AND region = (SELECT region FROM aws_regions WHERE region = 'us-east-2')));

INSERT INTO target_group (target_group_name, target_type, target_group_arn, ip_address_type, protocol, port, health_check_protocol, health_check_port, health_check_enabled, health_check_interval_seconds, health_check_timeout_seconds, healthy_threshold_count, unhealthy_threshold_count, health_check_path, protocol_version, region, vpc)
VALUES ('test', 'instance', 'arn:aws:elasticloadbalancing:us-east-2:257682470237:targetgroup/test/f699ce47a413e7e1', 'ipv4', 'HTTP', '80', 'HTTP', 'traffic-port', 't', '30', '5', '5', '2', '/', 'HTTP1', (SELECT region FROM aws_regions WHERE region = 'us-east-2'), (SELECT id FROM vpc WHERE vpc_id = 'vpc-024bfb31f9340a7cc' AND region = (SELECT region FROM aws_regions WHERE region = 'us-east-2')));

INSERT INTO listener (listener_arn, port, protocol, action_type, load_balancer_id, target_group_id)
VALUES ('arn:aws:elasticloadbalancing:us-east-2:257682470237:listener/app/test/88a5af6043314c2b/ec1b9847eae71ce9', '80', 'HTTP', 'forward', (SELECT id FROM load_balancer WHERE load_balancer_arn = 'arn:aws:elasticloadbalancing:us-east-2:257682470237:loadbalancer/app/test/88a5af6043314c2b'), (SELECT id FROM target_group WHERE target_group_arn = 'arn:aws:elasticloadbalancing:us-east-2:257682470237:targetgroup/test/f699ce47a413e7e1'));
```

### CODE REVIEW

**Comment by dfellis** [comment]
This error looks legit: https://github.com/iasql/iasql-engine/actions/runs/3942477857/jobs/6746196237#step:5:680

**Comment by aguillenv** [comment]
Re-opening PR addressing comments from https://github.com/iasql/iasql-engine/pull/1752

File: src/modules/iasql_functions/rpcs/iasql_get_sql_since.ts [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+  // For this we need to get Typeorm metadata for the `tableName` and inspect the columns and relations in order to recreate the sub-query if necessary.
+  // We need to recreate the sub-query because related columns might not be the same across databases connected to the same cloud account.
+  const metadata = await getMetadata(tableName, modsIndexedByTable, orm);
+  let columnMetadata: ColumnMetadata | undefined;
+  if (value && metadata) {
+    // If `metadata instanceof EntityMetadata` means that there's an Entity in Typeorm which it's table name is `tableName`
+    if (metadata instanceof EntityMetadata) {
+      columnMetadata = metadata.ownColumns
+        .filter(oc => oc.databaseName === key)
+        .map(oc => oc)
```
Comment (dfellis) [inline src/modules/iasql_functions/rpcs/iasql_get_sql_since.ts:232]: What does this `.map` do?

File: src/modules/iasql_functions/rpcs/iasql_get_sql_since.ts [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+      }
+    }
+    // If `metadata instanceof RelationMetadata` means that there's no Entity in Typeorm which it's table name is `tableName`,
+    // but theres a join table linking entities and `tableName` is that join table. In this case we need to check `joinColumns`
+    // which will have the columns from the owner of the relationship and `inverseJoinColumns` will have the columns coming from
+    // the other entities in the relationship.
+    if (metadata instanceof RelationMetadata) {
+      columnMetadata = metadata.joinColumns
+        .filter(jc => jc.databaseName === key)
+        .map(jc => jc)
```
Comment (dfellis) [inline src/modules/iasql_functions/rpcs/iasql_get_sql_since.ts:251]: Same here. These `.map` calls don't seem to do anything.

File: src/modules/iasql_functions/rpcs/iasql_get_sql_since.ts [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+          columnMetadata.referencedColumn?.databaseName ?? 'unknown_key',
+          value,
+          columnMetadata.relationMetadata?.entityMetadata,
+          modsIndexedByTable,
+          orm,
+        );
+      }
+      columnMetadata = metadata.inverseJoinColumns
+        .filter(jc => jc.databaseName === key)
+        .map(jc => jc)
```
Comment (dfellis) [inline src/modules/iasql_functions/rpcs/iasql_get_sql_since.ts:264]: and another one

File: src/modules/iasql_functions/rpcs/iasql_get_sql_since.ts (hunk 1 above)
Comment (aguillenv) [reply src/modules/iasql_functions/rpcs/iasql_get_sql_since.ts:232]: Alt Text

Comment (aguillenv) [reply src/modules/iasql_functions/rpcs/iasql_get_sql_since.ts:232]: Nothing, actually. Before I was getting the relation but then I decided to keep the column metadata for the final array check.

**Review by dfellis — APPROVED** [review]

**Review by depombo — APPROVED** [review]

**Pull request merged**